### DISCUSSION TITLE
Delete certain data from collection

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13576] [Q&A] [answered]
Want to `remove` certain models from my collection or `reset` collection. But after `doClick` the error log show all models. AFAIK the only way to delete data is clearing the whole collection via `Alloy.Collections.random.clear()`. How to delete certain models?
views/index.xml
```
<Alloy>
	<Window class="container" onClose="cleanup">
		<Label id="label" onClick="doClick">Hello, World</Label>
	</Window>
</Alloy>
```

models/random.js
```
exports.definition = {
    config: {
        "columns": {
            "name": "TEXT",
            "number": "NUMBER",
            "id": "INTEGER PRIMARY KEY AUTOINCREMENT"
        },
        "adapter": {
            "type": "sql",
            "collection_name": "random",
            "idAttribute": "id"
        }
    },
    extendModel: function(Model) {
      _.extend(Model.prototype, {
        constructor : function ( obj ) {
          Alloy.Model.apply(this, {number: obj.number,
                                   name: obj.name});
        }
      });
      return Model;
    },
    extendCollection: function(Collection) {
      _.extend(Collection.prototype, { // Extend, override or implement Backbone.Collection
          clear: function() {
            _.invoke(this.toArray(), 'destroy');
          }
      });

      return Collection;
    }
}

```

controllers/index.js
```
Alloy.Collections.instance('random');
record();
$.index.open();

function  record(){
for (let i = 0; i < 5; i++) {
    let no = Math.floor(Math.random() * 10) + 1;
    let model = Alloy.createModel('random', {name : "test" + no, number: no});
    model.save();
    Alloy.Collections.random.fetch();
    Ti.API.error(JSON.stringify(model));
  }
}

function doClick(e) {
  //let models = Alloy.Collections.random.get(1);
  //Alloy.Collections.random.remove(models);
  Alloy.Collections.random.reset();
  Alloy.Collections.random.fetch();
  Ti.API.error(JSON.stringify(Alloy.Collections.random.models));
}
function cleanup() {
    $.destroy();
}

```

**Comment by m1ga** [comment, marked as answer]
in my old demo I've did it like this;
one item:
https://github.com/m1ga/ti.model_collection_example/blob/master/app/controllers/basicExample.js#L34-L37

all items:
https://github.com/m1ga/ti.model_collection_example/blob/master/app/models/library.js#L39-L42

**Reply by beppo-ivel** [reply]
If I got the point, the problem is that `remove` or `reset` does not effects the persistence layer. So I have to sync my collection after remove or reset models, before I call tthe next `fetch`?

**Reply by m1ga** [reply]
not that deep in the backbone stuff :smile: But you can look at the methods in https://backbonejs.org/#Model-destroy (and in collection for reset, remove). Maybe that answers your question.

**Reply by beppo-ivel** [reply]
There is also a chapter in the titanium docs about how [Alloy](https://titaniumsdk.com/guide/Alloy_Framework/Alloy_Guide/Alloy_Models/Alloy_Sync_Adapters_and_Migrations.html) allows you to store and load models to a persistent storage. Backbone syncs models to persistent storage devices based on the implementation of the Backbone.sync method. The sync method depends on calls to other Backbone methods e.g. model.save(). Where the Alloy analogue of my example is:

```
let model = Alloy.createModel('random', {name : "test" + no, number: no});
model.save();
````
In the chapter linked above, there is also a table where Backbone method Collection.create (id != null) is referenced, which is equivalent to SQL method `UPDATE`. What is the Alloy analogue of this method? in my example above?

**Reply by m1ga** [reply]
again: not an backbone expert (or even heavy user) but perhaps `Alloy.createCollection`? Have to say that I only use it to store and read data. Nothing special. No custom sync adapter or so.

For more complex stuff I just use the database and run sql commands on it :smile:  `var db = Ti.Database.open('_alloy_');`

**Reply by beppo-ivel** [reply]
Where to find declaration of `Alloy.createCollection` to see implementation of Backbone.sync method for Collection.fetch, Collection.create, Model.fetch, Model.save and Model.destroy etc?

**Reply by m1ga** [reply]
https://github.com/tidev/alloy/blob/master/docs/apidoc/alloy.js#L58-L69
https://github.com/tidev/alloy/blob/5e9e25e58ce0fac3e8c12e7583297a5f64d27101/Alloy/lib/alloy/widget.js#L32-L34

**Reply by beppo-ivel** [reply]
That is not exactly what I'm looking for. This functions cannot persist to external storage without implementing the Backbone.sync method. Where is the default sql adapter definition (fetch, create, save etc.) for Alloy Models and Collections which allows to persist objects to external storage. I mean I already use SQL adapter to sync (e.g. `model.save()`) with the SQLite engine on Android.  

For Example: the custom `clear` function that you linked above, looks like a call of [Alloy Sync adapter](https://github.com/tidev/alloy/blob/5e9e25e58ce0fac3e8c12e7583297a5f64d27101/Alloy/lib/alloy/sync/properties.js#L42). And IMO the Alloy `destroy` function uses the same method. But where I can find the declaration of destroy, save, etc?

**Reply by m1ga** [reply]
https://github.com/tidev/alloy/tree/master/Alloy/lib/alloy/sync
just search in the Alloy repo, I'm doing the same :smile:

**Reply by beppo-ivel** [reply]
OK, where can I find the declaration of [setObject](https://github.com/tidev/alloy/blob/5e9e25e58ce0fac3e8c12e7583297a5f64d27101/Alloy/lib/alloy/sync/properties.js#L40)?

**Reply by m1ga** [reply]
Since that is `Ti.App.Properties` it is in the SDK and you'll find it here: https://github.com/tidev/titanium_mobile/commit/eed6159d69034f695d7eb21caebe45e478dbe641
and for Android it will end up in a loop of setStrings so this is the java part https://github.com/tidev/titanium_mobile/blob/f80e1dc35b7e1d8795fd096db2b350fddc94e88c/android/modules/app/src/java/ti/modules/titanium/app/properties/PropertiesModule.java#L114

edit: the master location of the first JS file is https://github.com/tidev/titanium_mobile/blob/master/common/Resources/ti.internal/extensions/ti/ti.app.properties.js

**Reply by beppo-ivel** [reply]
> just search in the Alloy repo, I'm doing the same 

Since this answer was given, I have made no new insights. In principal: If you want to r/w collection or model to external storage, then you need a sync adapter. Alloy comes with a [sql sync adapter](https://github.com/tidev/alloy/blob/5e9e25e58ce0fac3e8c12e7583297a5f64d27101/Alloy/lib/alloy/sync/sql.js). In general a sync adapter contains methods for read, create, update and delete data from external storage. According to the documentation of backbone and titanium as well, this methods are executed by the function calls collection.fetch(), collection.create(), model.fetch(), model.save() and model.destroy(). However I don't know which function to call to trigger the collection update. Even don't know if a update procedure is implemented for collection.

> For more complex stuff I just use the database and run sql commands on it smile var db = Ti.Database.open('_alloy_');

The sync adapter it self uses this mechanism. But I think I got your point, also [this project](https://github.com/FokkeZB/Tracker/blob/master/app/lib/DB.js) uses some helpers to do complex SQL queries and bypassing Backbone. Is there no update function implemented for collections in alloy?

**Reply by m1ga** [reply]
if backbone has it you can use it! Alloy is just using https://github.com/tidev/alloy/blob/master/Alloy/lib/alloy/backbone/1.4.0/backbone.js 
so I'm a bit confused what you are search for :smile: If you found a way in Backbone it should be the same in Alloy. But as I mentioned before: I don't use it so I'll stop here. I'm not really helping, just guessing. Perhaps someone else knows a bit more about it.
